src/coreComponents/constitutive/fluid/multifluid/compositional/CompositionalMultiphaseFluidUpdates.hpp

```diff
@@ -0,0 +1,431 @@
+  GEOS_UNUSED_VAR( pressure, temperature );
+
+  integer constexpr maxNumComp = MultiFluidBase::MAX_NUM_COMPONENTS;
+  integer constexpr maxNumPhase = MultiFluidBase::MAX_NUM_PHASES;
+  integer const numComp = numComponents();
+  integer const numPhase = numPhases();
+
+  // 1. Convert input mass fractions to mole fractions and keep derivatives
+
+  stackArray1d< real64, maxNumComp > compMoleFrac( numComp );
+  if( m_useMass )
+  {
+    convertToMoleFractions< maxNumComp >( composition, compMoleFrac );
+  }
+  else
+  {
+    for( integer ic = 0; ic < numComp; ++ic )
+    {
+      compMoleFrac[ic] = composition[ic];
+    }
+  }
+
+  // 2. Compute phase fractions and phase component fractions
+  m_flash.compute( pressure,
+                   temperature,
+                   compMoleFrac.toSliceConst(),
+                   phaseFrac,
+                   phaseCompFrac );
+
+  // 3. Calculate the phase densities
+  m_phase1.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[0].toSliceConst(),
+                            phaseDens[0],
+                            phaseMassDens[0],
+                            m_useMass );
+  m_phase2.density.compute( pressure,
+                            temperature,
+                            phaseCompFrac[1].toSliceConst(),
+                            phaseDens[1],
+                            phaseMassDens[1],
+                            m_useMass );
+  if constexpr (2 < FLASH::getNumberOfPhases())
+  {
+    m_phase3.density.compute( pressure,
+                              temperature,
+                              phaseCompFrac[2].toSliceConst(),
+                              phaseDens[2],
+                              phaseMassDens[2],
+                              m_useMass );
+  }
+
+  // 4. Calculate the phase viscosities
+  m_phase1.viscosity.compute( pressure,
+                              temperature,
+                              phaseCompFrac[0].toSliceConst(),
+                              phaseVisc[0],
+                              m_useMass );
+  m_phase2.viscosity.compute( pressure,
+                              temperature,
+                              phaseCompFrac[1].toSliceConst(),
+                              phaseVisc[1],
+                              m_useMass );
+  if constexpr (2 < FLASH::getNumberOfPhases())
+  {
+    m_phase3.viscosity.compute( pressure,
+                                temperature,
+                                phaseCompFrac[2].toSliceConst(),
+                                phaseVisc[2],
+                                m_useMass );
+  }
+
+  // 5. if mass variables used instead of molar, perform the conversion
+
+  if( m_useMass )
+  {
+
+    // unfortunately here, we have to copy the molecular weight coming from PVT package...
+    real64 phaseMolecularWeight[maxNumPhase]{};
+    for( integer ip = 0; ip < numPhase; ++ip )
+    {
+      phaseMolecularWeight[ip] = 1.0;
```

Comment:
Is it okay that 1.0 is assigned here, not the phase MW?  Just checking.

Reply:
Definitely not. All calculations are hard coded constants. Initial PR is to generate the object and kernels that will run. Subsequent PRs will link to actual calculation functions.